src/Microsoft.ML.Data/Training/TrainContext.cs

```diff
@@ -16,13 +22,136 @@ public abstract class TrainContextBase
+            int numFolds, string stratificationColumn)
+        {
+            Host.CheckValue(data, nameof(data));
+            Host.CheckValue(estimator, nameof(estimator));
+            Host.CheckParam(numFolds > 1, nameof(numFolds), "Must be more than 1");
+            Host.CheckValueOrNull(stratificationColumn);
+
+            EnsureStratificationColumn(ref data, ref stratificationColumn);
+
+            Func<int, (IDataView scores, ITransformer model)> foldFunction =
+                fold =>
+                {
+                    var trainFilter = new RangeFilter(Host, new RangeFilter.Arguments
+                    {
+                        Column = stratificationColumn,
+                        Min = (double)fold / numFolds,
+                        Max = (double)(fold + 1) / numFolds,
+                        Complement = true
+                    }, data);
+                    var testFilter = new RangeFilter(Host, new RangeFilter.Arguments
+                    {
+                        Column = stratificationColumn,
+                        Min = (double)fold / numFolds,
+                        Max = (double)(fold + 1) / numFolds,
+                        Complement = false
+                    }, data);
+
+                    var model = estimator.Fit(trainFilter);
+                    var scoredTest = model.Transform(testFilter);
+                    return (scoredTest, model);
+                };
+
+            // Sequential per-fold training.
+            // REVIEW: we could have a parallel implementation here. We would need to
+            // spawn off a separate host per fold in that case.
+            var result = new List<(IDataView scores, ITransformer model)>();
+            for (int fold = 0; fold < numFolds; fold++)
+                result.Add(foldFunction(fold));
+
+            return result.ToArray();
+        }
+
         protected TrainContextBase(IHostEnvironment env, string registrationName)
         {
             Contracts.CheckValue(env, nameof(env));
             env.CheckNonEmpty(registrationName, nameof(registrationName));
             Host = env.Register(registrationName);
         }
 
+        /// <summary>
+        /// Make sure the provided <paramref name="stratificationColumn"/> is valid
+        /// for <see cref="RangeFilter"/>, hash it if needed, or introduce a new one
+        /// if needed.
+        /// </summary>
+        private void EnsureStratificationColumn(ref IDataView data, ref string stratificationColumn)
+        {
+            // We need to handle two cases: if the stratification column is provided, we use hashJoin to
+            // build a single hash of it. If it is not, we generate a random number.
+
+            if (stratificationColumn == null)
+            {
+                stratificationColumn = data.Schema.GetTempColumnName("StratificationColumn");
```

Comment:
>StratificationColumn [](start = 70, length = 20)

is there a DefaultColumnName for this?

Reply:
I looked, there is none. I don't think there should be either, it's a pretty peculiar concept, unique to CV and split scenarios

---
In reply to: [220372046](https://github.com/dotnet/machinelearning/pull/1005#discussion_r220372046) [](ancestors = 220372046)